README.md
# NBE: Node-Based Editing

A node-based-editing control based on web technologies

NBE is a library to help develop node-based-editing applications using AngularJS.
It is implemented as a set of AngularJS _directives_ which collaborate to implement a node-based editor.
Links between vertices and edges are rendered using SVG, so only fairly recent browsers are supported:
Chrome and Firefox, Safari 7.1+ and MSIE 11+ should work decently well.
Applications speak to NBE by binding its graph model (see below) and potentially a layout and edge-type model (see below).


## The Model

The model of NBE is a directed (hyper-)graph consisting of vertices and typed edges:

- Each vertex may have an arbitrary number of labeled and typed input and output ports.
- Each port may be connected to 0..1 edges of matching type.
- Each edge may be connected to any number of inputs or outputs (1:n, n:m or n:1 depending on the edge type).


## Current Status

The project is still relatively immature, but already has a couple of nice features that might justify having a closer look.
New features and bugfixes are tracked in the [CHANGELOG](CHANGELOG.md).

#### What NBE offers
- display and manipulate directed (hyper-)graph models
- undo/redo structural operations using the keyboard
- select and cut/copy/paste graph parts
- changes to the model (triggered by the host application) are observed by the editor
- automatic layout using the [dagre library](https://github.com/cpettitt/dagre)
- edges for a given type are either simple edges (1:n, n:1) or complex _hyperedges_ (n:m)
- vertex types, to destinguish between different classes of vertices
- hyperedges have multiple incoming and outgoing links, and are represented by circles
- zoom in and out
- customize the appearance of the editor using (S)CSS

#### Planned Features
- undo/redo of layout operations
- specs and spec-tests for the graph controller operations which are exposed as an API
- [more planned features](https://github.com/x1B/nbe/issues?q=is%3Aopen+is%3Aissue+label%3Aenhancement)

#### Known Bugs
_no bugs on record at time of release_
- [more bugs](https://github.com/x1B/nbe/issues?q=is%3Aopen+is%3Aissue+label%3Abug)


## Getting Started

Included are several examples showing how to integrate NBE into projects which are using _[Bower](http://bower.io/)_ and _[RequireJS](http://requirejs.org/)_.  

### Setup

The examples also contain a _Gruntfile_ to build an optimized version, which is completely optional. 
To get started, run the `logic` example app, which allows you to edit a logical circuit and to simulate its operation:

```sh
# get the sources
git clone https://github.com/x1B/nbe.git
# set up dependencies, shared by all examples
bower install
```

Serve the directory where you extracted nbe from any web server.
Navigate to `examples/logic/debug.html` (using Mozilla Firefox or Google Chrome for now) to launch the demo application.
Or browse to `examples/logic/index.html` for the optimized version.    
If you make changes, the optimized version can be rebuilt (requires `npm` and `grunt`):
```sh
cd examples/logic
npm install
```

### Using the Node-Based Editor

Try adding logic signals and gates to the demo circuit, and connect them by dragging wires between the nodes.
Hit _'Flatten Main Circuit'_ to expand circuit compositions within the graph into a flattened model.
Use _'Run Simulation'_ to view state changes of the probed wires as the discrete clock drives the simulation.
The boxes represent vertices in the hypergraph, and the curved lines (_link_) make up the edges.
Hit _'Auto Layout'_ to tidy up the graph.
In this example, each _wire_ edge may have one source and multiple outgoing links (1:n), while log _channels_ are n:m hyperedges, which may (for the sake of demonstration) have multiple sinks.

You can double click on a connected _port_ (the socket where links connect to the vertices), to sever its link(s).
Also, you can select one or more nodes by clicking or by dragging a box to move or auto-layout them as a group, or hit _del_ to remove them.
Basic undo/redo functionality is provided by the usual shortcuts.

### The API

Here is a quick walk-through of the API based on the `logic` example application:

#### The Graph

The graph model is a JavaScript object containing two maps, `edges` and `vertices`.
As the example shows (entry `main` in file `lib/data/dummy_model.json`), edges and vertices are stored under their respective IDs, occupying separate namespaces.
Edges have the string properties `type` and `label`, while vertices have a `label` as well as two lists of _ports_ (`in` and `out`).
Each port is an object with a `label`, an `id` (unique within the vertex), an optional `direction` which is either `"out"` or `"in`" (default), and an optional `edgeId` referencing the edge a port is connected to (if any). 

Applications instantiate a node-based editor using the AngularJS directive `nbeGraph`, which should be referenced as `data-nbe-graph`.
Have a look at the file `lib/logic_demo.html` for an example of an NBE embedding.
For styling to work, it is recommended to put the `nbeGraph` onto a DIV, and to put that DIV inside another block-level element with fixed dimensions and CSS class `graph-container`.
The directive takes an AngularJS binding expression that must evaluate to a graph-model as described above.

#### Edge Types

Use the binding attribute `data-nbe-graph-types` on the graph element to configure the edge types of your application.
Each edge type gets its own color, to be defined in the CSS files.
It is strongly recommended to use `compass` to generate edge-type CSS for a given application as illustrated in the example under `scss/`.
The edge types are configured as an object: 
Each edge types has a name (the key) and up to three properties as the value object:
 * the property `simple` (boolean) determines if an edge is just a 1:n (or n:1) edge (`true`), or an n:m _hyperedge_ (`false`, default)
 * the property `maxSources` (number) puts a maximum on the number of incoming links for each edge
 * the property `maxDestinations` (number) puts a maximum on the number of outgoing links for each edge

#### The Layout

Applications may access and modify the graph layout, for example to persist it to local storage or onto a server.
It must be bound to the `data-nbe-graph-layout` attribute on the graph element.
The layout model looks similar to the data model, but contains `left` and `top` pixel coordinates for each node.
Have a look at the file `lib/data/dummy_layout.json` and you will get the idea.

#### Binding to the graph controller API

Any UI that is not directly associated with individual graph elements (nodes, edges, links) must be implemented by the application.
The logic demo has several buttons that are bound to the provided controller actions:
 * `calculateLayout()` invokes the Dagre layout algorithm on the selected nodes (or all nodes if the selection is empty)
 * `zoom.zoomIn()` and `zoom.zoomOut()` change the scale of the graph which is useful for large graphs


## Feedback

Please let [the original author](http://x1b.github.com) know if you have built something using NBE, or if you encounter any problems or questions.


CHANGELOG.md
# Changelog

## Last Changes

- [#50](https://github.com/x1B/nbe/issues/50): operations: allow graph model to be readonly
    * _New feature:_ see ticket for details

- [#44](https://github.com/x1B/nbe/issues/44): performance: allow to trigger graph updates through AngularJS events
    * _New feature:_ see ticket for details

- [#66](https://github.com/x1B/nbe/issues/66): layout: fixed glitch before initial layout

- [#64](https://github.com/x1B/nbe/issues/64): updates: fixed cleanup of link view model

- [#63](https://github.com/x1B/nbe/issues/63): selection: fixed clearing by click on canvas (firefox)

- [#62](https://github.com/x1B/nbe/issues/62): refactored and simplified CSS classes, added prefix `nbe`

- [#60](https://github.com/x1B/nbe/issues/60): drag/drop: make it easier to connect ports

- [#61](https://github.com/x1B/nbe/issues/61): selection: fixed highlighting implicitly selected edges

- [#59](https://github.com/x1B/nbe/issues/59): selection: fixed style priority

- [#58](https://github.com/x1B/nbe/issues/58): selection: fixed flickering in safari and positioning in firefox


## v0.4.0

- [#55](https://github.com/x1B/nbe/issues/55): prepared codebase for release

- [#52](https://github.com/x1B/nbe/issues/52): refactoring: compatibility with to AngularJS 1.3.x


## v0.3.0

- [#49](https://github.com/x1B/nbe/issues/49): footprint: removed underscore


## v0.2.0

- [#48](https://github.com/x1B/nbe/issues/48): release: prepared documentation and logic demo for v0.2.0

- [#15](https://github.com/x1B/nbe/issues/15): SCSS: removed vendor prefixes, added auto-prefixer to examples

- [#47](https://github.com/x1B/nbe/issues/47): utilities/traverse: added missing check to `eachConnectedPort`

- [#43](https://github.com/x1B/nbe/issues/43): allow to cut/copy/paste nodes and edges from/to the current graph
    * users can now focus one graph at a time (using the tab key or mouse clicks)
    * using Ctrl-C/Ctrl-X the current selection of nodes (if any) is copied/moved into a buffer
    * if supported by the browser, a JSON representation of the selection is put into the global clipboard
    * using Ctrl-V, the buffer may be pasted into the currently focused graph (which may be different from source)
    * the nodes that were just pasted become the new graph selection
    * links between selected nodes are retained when moving to the clipboard, connections to other nodes are lost 

- [#45](https://github.com/x1B/nbe/issues/45): fixed removal of vertices from the graph model

- [#31](https://github.com/x1B/nbe/issues/31): allow to classify nodes using "classes" attribute

- [#42](https://github.com/x1B/nbe/issues/42): partition ports into inbound and outbound subset

- [#41](https://github.com/x1B/nbe/issues/41): fixed layout of simple edges in selection


## v0.1.0

- [#40](https://github.com/x1B/nbe/issues/40): prepared codebase for release v0.1.0

- [#39](https://github.com/x1B/nbe/issues/39): fixed port dragging and double click

- [#38](https://github.com/x1B/nbe/issues/38): refactored the graph directive controller

- [#37](https://github.com/x1B/nbe/issues/37): implemented configurable zoom levels based on SCSS mixins

- [#36](https://github.com/x1B/nbe/issues/36): added 'simple' edge types for 1:n and n:1 relations

- [#33](https://github.com/x1B/nbe/issues/33): added informations to install app

- [#17](https://github.com/x1B/nbe/issues/17): added several (incomplete) examples, updated changelog

- [#32](https://github.com/x1B/nbe/issues/32): fixed dragging of diconnected nodes

- [#22](https://github.com/x1B/nbe/issues/22): updated code-style (no space before opening brace)

- [#29](https://github.com/x1B/nbe/issues/29): implemented multiple selection:
    * select multiple nodes by dragging a rectangle
    * delete the complete selection at once
    * auto-layout only applies to selected nodes (or all nodes)
    * dragging a selected node drags the entire selection

- [#28](https://github.com/x1B/nbe/issues/28): implemented circuit simulator demo

- [#26](https://github.com/x1B/nbe/issues/26): added a short ping animation which is triggered when a link is formed

- [#25](https://github.com/x1B/nbe/issues/25): fix link pathing during port drag/drop

- [#24](https://github.com/x1B/nbe/issues/24): disconnect ports using doubleclick instead of click

- [#23](https://github.com/x1B/nbe/issues/23): injected more things
    * extracted constant values into injectable settings
    * replaced async utility with async service
    * added async service

- [#12](https://github.com/x1B/nbe/issues/12): allowed to filter visible edges by type
    * ports of a hidden type are hidden as well
    * only visible edges are included in automatic layout

- [#11](https://github.com/x1B/nbe/issues/11): allowed to restrict edge cardinality by type and port direction (source ports, dest ports)

- [#21](https://github.com/x1B/nbe/issues/21): fixed direction-dependent highlighting of ports on drag

- [#20](https://github.com/x1B/nbe/issues/20): allowed graph inside scroll-container to grow, allowed multiple graphs on the same page, to be used by different AngularJS modules.

- [#18](https://github.com/x1B/nbe/issues/18): allowed to provide a button for automatic layout

- [#9](https://github.com/x1B/nbe/issues/9): fixed placement of edges upon creation

- [#16](https://github.com/x1B/nbe/issues/16): made it simpler to disconnect ports from edges: just click on the port

- [#3](https://github.com/x1B/nbe/issues/3): implemented automatic layout calculation using Dagre

- [#14](https://github.com/x1B/nbe/issues/14): fixed port dragging after refactoring

- [#8](https://github.com/x1B/nbe/issues/8): refactoring: moved operations (undo/redo) to a new module

- [#6](https://github.com/x1B/nbe/issues/6): moved graph HTML into the directives

- [#7](https://github.com/x1B/nbe/issues/7): cleaned up most dead code


js/dummy_data.js
define([], function () {

   var data = {
      "dummyModel": {
         vertices: {
            "v0": {
               label: "A Resource Master",
               ports: {
                  "in": {
                     "i0": { label: "Enable On", type: "FLAG" },
                     "i1": { label: "My Trigger", type: "ACTION" }
                  },
                  "out": {
                     "o0": { label: "Resource (M)", type: "RESOURCE" },
                     "o1": { label: "Confirm", type: "ACTION", edge: "e1" }
                  }
               }
            },
            "v1": {
               label: "Weird Stuff",
               ports: {
                  "in": {
                     "i0": { label: "Some Resource (S)", type: "RESOURCE" },
                     "i1": { label: "Another Resource (S)", type: "RESOURCE" },
                     "i2": { label: "Play Music On", type: "FLAG" }
                  },
                  "out": {
                     "o0": { label: "Other Resource (M)", type: "RESOURCE", edge: "e0" },
                     "o1": { label: "Cancel Action", type: "ACTION" }
                  }
               }
            },
            "v2": {
               label: "Silly Stuff",
               ports: {
                  "in": {
                     "i0": { label: "The Resource (S)", type: "RESOURCE", edge: "e0" },
                     "i1": { label: "Shuffle Action", type: "ACTION" }
                  },
                  "out": {
                     "o0": { label: "Another Resource (M)", type: "RESOURCE" },
                     "o1": { label: "Refresh Action", type: "ACTION" }
                  }
               }
            },
            "v3": {
               label: "A Big One",
               ports: {
                  "in": {
                     "i0": { label: "First Resource (S)", type: "RESOURCE", edge: "e0" },
                     "i1": { label: "Second Resource (S)", type: "RESOURCE" },
                     "i2": { label: "Take a Second", type: "FLAG" },
                     "i3": { label: "Save Action", type: "ACTION", edge: "e1" }
                  },
                  "out": {
                     "o0": { label: "Third Resource (M)", type: "RESOURCE" },
                     "o1": { label: "Fourth Resource (M)", type: "RESOURCE" },
                     "o2": { label: "Use the Fourth", type: "FLAG" }
                  }
               }
            }
         },
         "edges": {
            "e0": {
               "label": "myRessource",
               "type": "RESOURCE"
            },
            "e1": {
               "label": "myAction",
               "type": "ACTION"
            },
            "e2": {
               "label": "myFlag",
               "type": "FLAG"
            }
         }
      },

      "dummyLayout": {
         "vertices": {
            "v0": { "left": "10%", "top": "10%" },
            "v1": { "left": "15%", "top": "60%" },
            "v2": { "left": "70%", "top": "20%" },
            "v3": { "left": "60%", "top": "50%" }
         },
         "edges": {
            "e0": { "left": "40%", "top": "30%" },
            "e1": { "left": "40%", "top": "60%" },
            "e2": { "left": "50%", "top": "70%" }
         }
      }
   };

   return {
      get: function ( key ) {
         return data[ key ];
      }
   }
});